docs: iter 4 (#532)
iterative/terraform-provider-iterative: README.md

TPI is a [Terraform](https://terraform.io) plugin built with machine learning in mind. This CLI tool offers full lifecycle management of computing resources (including GPUs and respawning spot instances) from several cloud vendors (AWS, Azure, GCP, K8s)... without needing to be a cloud expert.

> - **Lower cost with spot recovery**: transparent data backup & auto-recovery from interrupted low-cost spot/preemptible instances

Reviewer: ```suggestion
- **Lower cost with spot recovery**: transparent data checkpoint & restore from interrupted low-cost spot/preemptible instances
```

Traditionally “backup” has a long–term connotation that may be misleading. 🤔 Not that my suggested change is much clearer `:muk:` but probably conveys better the functionality.

Author: ```suggestion
- **Lower cost with spot recovery**: transparent data checkpoint/restore & auto-respawning of low-cost spot/preemptible instances
```

   TPI is a CLI tool, not a running service. It requires no additional orchestrating machine (control plane/head nodes) to schedule/recover/terminate instances. Instead, TPI runs (spot) instances via cloud-native scaling groups[^scalers], taking care of recovery and termination automatically on the cloud provider's side. This design reduces management overhead & infrastructure costs. You can close your laptop while cloud tasks are running -- auto-recovery happens even if you are offline.
2. **Unified tool for data science and software development teams**:
   TPI provides consistent tooling for both data scientists and DevOps engineers, improving cross-team collaboration. This simplifies compute management to a single config file, and reduces time to deliver ML models into production.
3. **Reproducible, codified environments**:

>    Store hardware requirements in a single configuration file alongside the rest of your ML pipeline code.

Reviewer 2: When we number the list it seems like we mean that it is an exhaustive list. Is it? Otherwise maybe bullets?

But also it somewhat overlaps with the previous bullet list. Maybe they can be combined?

>    Store hardware requirements & pipelines in a single configuration file with the rest of your ML project code.

Reviewer: ```suggestion
   Store hardware requirements in a single configuration file with the rest of your ML project code.
```

As per https://github.com/iterative/terraform-provider-iterative/issues/513#issuecomment-1102383582, the “pipelines” part might be a bit far–fetched.

Author: how about this?

```suggestion
   Store hardware requirements in a single configuration file with the rest of your ML pipeline code.
```

Reviewer: Sounds better, but still implies (?) that users should store all their pipeline code in a Terraform file. 🙃

Author: Hardware requirements (main.tf) are stored with/alongside ML pipeline code (dvc.yaml, train.py) in Git.

Reviewer: ```suggestion
   Store hardware requirements & pipelines in a single configuration file, alongside with the rest of your ML pipeline code.
```

resource "iterative_task" "example" {
  cloud      = "aws" # or any of: gcp, az, k8s
  machine    = "m"   # medium. Or any of: l, xl, m+k80, xl+v100, ...

>   spot        = 0    # auto-price. Default -1 to disable, or >0 for hourly USD limit

Reviewer: ```suggestion
  spot       = 0    # auto-price. Default -1 to disable, or >0 for hourly USD limit
```

I'm the incarnation of `terraform fmt` 🐉

  disk_size  = 30    # GB

  storage {
    workdir = "."       # default blank (don't upload)

>     output  = "results" # default blank (don't upload). Relative to workdir

Reviewer: ```suggestion
    output  = "results" # default blank (don't download). Relative to workdir
```

TF_LOG_PROVIDER=INFO terraform show
```

> ### End Task

Reviewer: ```suggestion
### Delete Task
```

More accurate?

Author: - stop: clashes with "stop instance" meaning
- delete: only applies to storage data, not instance
- terminate: only applies to instance, not storage data
- destroy: accurate Terraform term, but sounds weird and also is a self-referential definition?
- finish/end: new term

I'd say finish/end/destroy in order of preference

Reviewer: > _delete_

Delete like in “deleting a cloud resource” and not like in “deleting a file”. Niche, but accurate enough. ⚔️

Reviewer: > _destroy_

Terraform–specific and a bit exagerated, but also pretty descriptive.

Reviewer: > _cleanup_

Reminder: it also downloads data

Reviewer: For the sake of conceptual creativity: the original XPD proposal used “harvest”

Author: leaving this unresolved until someone comes up with any better ideas :)

Reviewer: 👍🏼
